### PR TITLE
feat: add design document for binst install command

### PR DESCRIPTION
**Pull request by haya14busa**
This PR adds a comprehensive design document for implementing the `binst install` command.

The design document covers:
- Architecture and package structure
- Identification of reusable existing packages
- Detailed design for new packages
- 5-phase implementation plan
- Testing strategy including parity tests
- Security considerations

This design ensures the `binst install` command will have script-parity with generated installers while creating a reusable foundation for future features.

Related to #141

Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: doc/design/binst-install-command.md [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+3. Additional verification methods beyond current installer capabilities
+4. UX improvements beyond script parity
+
+## Architecture
+
+### Package Structure
+
+```
+pkg/
+├── config/          # Config loading and discovery
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:27]: Do we really need this package considering we have `cmd/shared.go`?

File: doc/design/binst-install-command.md [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+├── resolve/         # Asset resolution logic
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:28]: Isn't `pkg/asset.FilenameGenerator` enough? I don't think we need another wrapper package.

File: doc/design/binst-install-command.md [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+// Download fetches a file to a temporary location
+func (d *Downloader) Download(url string) (string, error)
+
+// DownloadWithRetry implements retry logic matching shell script
+func (d *Downloader) DownloadWithRetry(url string, maxRetries int) (string, error)
+```
+
+**Implementation Notes:**
+- Use `pkg/httpclient` for GitHub API access
+- Implement same retry logic as shell script (3 retries by default)
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:114]: No need to support retry.

File: doc/design/binst-install-command.md [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+## Architecture
+
+### Package Structure
+
+```
+pkg/
+├── config/          # Config loading and discovery
+├── resolve/         # Asset resolution logic
+├── fetch/           # Download with retry logic
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:29]: I don't think we need this package. `pkg/httpclient`should be enough.

File: doc/design/binst-install-command.md [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+- Each package gets comprehensive unit tests
+- Mock external dependencies (GitHub API, filesystem)
+- Test error conditions and edge cases
+
+### Integration Tests
+- Test full installation flow
+- Use testdata configurations
+- Compare results with generated installers
+
+### Parity Tests
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:252]: We can have e2e test with scripts and make task instead.

File: doc/design/binst-install-command.md [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+## Future Extensibility
+
+This design enables future `binst x` command by:
+1. Reusing resolution/fetch/verify packages
+2. Adding caching layer on top
+3. Executing in temporary directory instead of installing
+
+## Open Questions
+
+1. Should we support parallel asset downloads for faster installation?
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:299]: No. We should download only one asset in the first place. 
Exception is checksum file.

File: doc/design/binst-install-command.md [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+2. How should we handle GitHub API rate limits?
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:300]: We can just support optional GitHub token.

File: doc/design/binst-install-command.md [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+3. Should dry-run mode make network requests?
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:301]: Yes, it's fine. We can just skip installation process for dry run.

**Review by haya14busa — COMMENTED** [review]
@claude Please address these review comments.

File: doc/design/binst-install-command.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+3. Additional verification methods beyond current installer capabilities
+4. UX improvements beyond script parity
+
+## Architecture
+
+### Package Structure
+
+```
+pkg/
+├── verify/          # Checksum verification
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:27]: We should use pkg/checksums for this use case.

File: doc/design/binst-install-command.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+### Phase 5: End-to-End Testing
+1. Create e2e test suite with scripts and make tasks
+2. Test on multiple platforms (Linux, macOS)
+3. Test with various projects from testdata/
+4. Ensure behavioral compatibility with generated scripts
+
+## Testing Strategy
+
+### Unit Tests
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:197]: Make sure not to use testify and use go-cmp for comparison if necessary.

**Review by haya14busa — COMMENTED** [review]
@claude Please address these review comments.

File: doc/design/binst-install-command.md [hunk 11]
```diff
@@ -198,6 +199,7 @@ func runInstall(cmd *cobra.Command, args []string) error {
 - Each package gets comprehensive unit tests
 - Mock external dependencies (GitHub API, filesystem)
 - Test error conditions and edge cases
+- Use go-cmp for comparison instead of testify
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:202]: I meant we shouldn't use testify in general.

File: doc/design/binst-install-command.md [hunk 12]
```diff
@@ -147,13 +147,15 @@ func runInstall(cmd *cobra.Command, args []string) error {
         return err
     }
 
-    // 3. Use pkg/asset.FilenameGenerator for asset resolution
-    // 4. Use pkg/httpclient.NewGitHubClient() for downloading
-    // 5. Verify, extract, and install
+    // 3. Resolve version (latest if not specified)
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:150]: We should create a package for this.

File: doc/design/binst-install-command.md [hunk 13]
```diff
@@ -24,7 +24,7 @@ This document outlines the design for implementing the `binst install` command t
 
 ```
 pkg/
-├── verify/          # Checksum verification
+├── verify/          # Checksum verification (uses pkg/checksums)
```
Comment (haya14busa) [inline doc/design/binst-install-command.md:27]: Do not create this package as discussed. 
pkg/checksums should be enough

**Review by haya14busa — COMMENTED** [review]
@claude Please address these review comments.
